app/src/main/java/uk/ac/ebi/atlas/search/species/SpeciesSearchDao.java

```diff
@@ -0,0 +1,122 @@
+        //      ),
+        //  on="bioentity_identifier"
+        // )
+
+        if (StringUtils.isBlank(searchText)) {
+            return Optional.of(ImmutableSet.of());
+        }
+
+        UniqueStreamBuilder uniqueBioEntitiesStreamBuilder = buildUniqueStreamBuilder(searchText, category);
+
+        UniqueStreamBuilder uniqueGene2ExperimentStreamBuilder = buildUniqueStreamBuilder();
+
+        var speciesStreamQueryBuilder = new InnerJoinStreamBuilder(
+                uniqueBioEntitiesStreamBuilder,
+                uniqueGene2ExperimentStreamBuilder,
+                BIOENTITY_IDENTIFIER.name()
+        );
+
+        return getSpeciesFromStreamQueryBuilder(speciesStreamQueryBuilder);
+    }
+
+    private Optional<ImmutableSet<String>> getSpeciesFromStreamQueryBuilder(InnerJoinStreamBuilder speciesStreamQueryBuilder) {
+        try (TupleStreamer tupleStreamer = TupleStreamer.of(speciesStreamQueryBuilder.build())) {
+            return Optional.of(tupleStreamer.get()
+                    .map(tuple -> tuple.getString(SPECIES.name()))
+                    .collect(toImmutableSet())
+            );
+        }
+    }
+
+    private UniqueStreamBuilder buildUniqueStreamBuilder() {
+        var gene2ExperimentQueryBuilder = new SolrQueryBuilder<Gene2ExperimentCollectionProxy>()
+                .setFieldList(Gene2ExperimentCollectionProxy.BIOENTITY_IDENTIFIER)
+                .sortBy(Gene2ExperimentCollectionProxy.BIOENTITY_IDENTIFIER, SolrQuery.ORDER.asc);
+        var searchGene2ExperimentBuilder =
+                new SearchStreamBuilder<>(gene2ExperimentCollectionProxy, gene2ExperimentQueryBuilder).returnAllDocs();
+        return new UniqueStreamBuilder(searchGene2ExperimentBuilder, BIOENTITY_IDENTIFIER.name());
+    }
+
+    private UniqueStreamBuilder buildUniqueStreamBuilder(String searchText, String category) {
+        var bioEntitiesByTextAndCategory = new SolrQueryBuilder<BioentitiesCollectionProxy>()
+                .addQueryFieldByTerm(PROPERTY_VALUE, searchText)
+                .setFieldList(ImmutableSet.of(SPECIES_DV, BIOENTITY_IDENTIFIER_DV))
+                .sortBy(BIOENTITY_IDENTIFIER_DV, SolrQuery.ORDER.asc);
+
+        if (StringUtils.isNotBlank(category)) {
+            bioEntitiesByTextAndCategory.addQueryFieldByTerm(PROPERTY_NAME, category);
+        }
+
+        var searchBioEntitiesBuilder =
+                new SearchStreamBuilder<>(bioentitiesCollectionProxy, bioEntitiesByTextAndCategory).returnAllDocs();
+        var selectBioEntitiesStreamBuilder =
+                new SelectStreamBuilder(searchBioEntitiesBuilder)
+                        .addFieldMapping(ImmutableMap.of(
+                                BIOENTITY_IDENTIFIER_DV.name(), BIOENTITY_IDENTIFIER.name(),
+                                SPECIES_DV.name(), SPECIES.name()
+                        ));
+        return new UniqueStreamBuilder(selectBioEntitiesStreamBuilder, BIOENTITY_IDENTIFIER.name());
+    }
```

Comment:
I’m not a big fan of throw-away methods or variables unless they introduce clarity of concepts in the code. For instance, `getSpeciesFromStreamBuilder` encapsulates under a name what it does, but `buildUniqueStreamBuilder` not so much. In the end is a matter of taste and I’ll leave it to your judgment. Just some food for thought.

Reply:
Yep. I also realized it after I pushed it to GH, but I thought I am going to rename these badly named methods when I do the change requests of the code review or in my next PR.

Reply:
Renamed a bunch of methods and variables. Hopefully, these names are more meaningful.